codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/customization/S3100Continue.java

```diff
@@ -0,0 +1,103 @@
+import software.amazon.smithy.go.codegen.SymbolUtils;
+import software.amazon.smithy.go.codegen.integration.ConfigField;
+import software.amazon.smithy.go.codegen.integration.GoIntegration;
+import software.amazon.smithy.go.codegen.integration.MiddlewareRegistrar;
+import software.amazon.smithy.go.codegen.integration.RuntimeClientPlugin;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.shapes.ServiceShape;
+import software.amazon.smithy.model.traits.HttpTrait;
+import software.amazon.smithy.utils.ListUtils;
+
+import java.util.List;
+
+/**
+ * Add middleware, which adds {Expect: 100-continue} header for s3 client HTTP PUT request larger than 2MB
+ * or with unknown size streaming bodies, during operation builder step
+ */
+public class S3100Continue implements GoIntegration {
+    private static final String ADD_100Continue_Header = "add100Continue";
+    private static final String ADD_100Continue_Header_INTERNAL = "Add100Continue";
+    private static final String ADD_100Continue_Header_Option = "AddContinueOption";
+    private static final String Continue_Client_Option = "ContinueHeaderThresholdBytes";
+
+    /**
+     * Return true if service is Amazon S3.
+     *
+     * @param model   is the generation model.
+     * @param service is the service shape being audited.
+     */
+    private static boolean isS3Service(Model model, ServiceShape service) {
+        return S3ModelUtils.isServiceS3(model, service);
+    }
+
+    /**
+     * Gets the sort order of the customization from -128 to 127, with lowest
+     * executed first.
+     *
+     * @return Returns the sort order, defaults to -40.
+     */
+    @Override
+    public byte getOrder() {
+        return 126;
+    }
+
+    @Override
+    public void writeAdditionalFiles(
+            GoSettings settings,
+            Model model,
+            SymbolProvider symbolProvider,
+            GoDelegator goDelegator
+    ) {
+        ServiceShape service = settings.getService(model);
+        if (!isS3Service(model, service)) {
+            return;
+        }
+
+        goDelegator.useShapeWriter(service, this::writeMiddlewareHelper);
+    }
+
+    private void writeMiddlewareHelper(GoWriter writer) {
+        writer.openBlock("func $L(stack *middleware.Stack, options Options) error {", "}", ADD_100Continue_Header, () -> {
+            writer.write("return $T(stack, options.ContinueHeaderThresholdBytes)",
+                    SymbolUtils.createValueSymbolBuilder(ADD_100Continue_Header_INTERNAL,
+                            AwsCustomGoDependency.S3_SHARED_CUSTOMIZATION).build()
+            );
+        });
+        writer.insertTrailingNewline();
+    }
+
+    @Override
+    public List<RuntimeClientPlugin> getClientPlugins() {
+        return ListUtils.of(
+                RuntimeClientPlugin.builder()
+                        .operationPredicate((model, service, operation) -> isS3Service(model, service) && operation.
```

Comment:
**fix**: We probably really only care about the `PutObject` and `UploadPart` operations. The other APIs don't have large binary payloads where an `Expect` header makes any difference (and may not be supported by the service for those operations IDK).

Reply:
**Question**: Can service and operation predicates be combined such that you can move the `isS3Service` check to `servicePredicate(...)`?

Reply:
Do you mean remove `isS3Service` by writing its logic directly within service predicate, I can do that. `isS3Service` is written just because it can be used in both `writeAdditionalFiles` and `getClientPlugins`.

Reply:
Yep, but I'm confused that some middleware codegen like [S3GetBucketLocation](https://github.com/aws/aws-sdk-go-v2/blob/main/codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/customization/S3GetBucketLocation.java#:~:text=.operationPredicate((model%2C%20service%2C%20operation)%20%2D%3E%20%7B) actually adds `isS3Service` check to operation predicate. Not sure which format is better